### ISSUE TITLE
TypeError: this.once is not a function

### ISSUE THREAD

**Issue by FreedCapybara** [mqttjs/MQTT.js#696]
Hello,

I'm getting `TypeError: this.once is not a function` on calling `mqtt.connect()`. I have looked at issues #597 and #596, but my current versions are mqtt@2.13.0, and readable-stream@2.3.3.

The project uses Angular with Webpack, and I get the same error whether I use ngx-mqtt or mqtt.js directly. Unfortunately I'm not sure how to reproduce the issue, since the library works as expected in a fresh project created with the Angular CLI (which also uses Webpack).

I can probably share my Webpack configuration if it would help.

Thank you!

**Comment by RangerMauve** [comment]
Could you please show the code you're using to invoke mqtt.connect()?

**Comment by FreedCapybara** [comment]
Here you go:

```js
import { Injectable } from '@angular/core';
import * as mqtt from 'mqtt';

@Injectable()
export class MqttService {

  private _client: any;

  constructor() {
    this._client = mqtt.connect('mqtt://iot.eclipse.org:80/ws');
    this._client.on('connect', function () {
      console.log('connected');
    });
  }
}
```

**Comment by RangerMauve** [comment]
Could you also please provide a stack trace for the error?

**Comment by FreedCapybara** [comment]
There's a lot of stack:
```
ERROR TypeError: this.once is not a function
    at BufferList.Duplex (_stream_duplex.js:70)
    at new BufferList (bl.js:32)
    at BufferList (bl.js:7)
    at Parser._resetState (parser.js:27)
    at new Parser (parser.js:19)
    at Object.Parser [as parser] (parser.js:10)
    at MqttClient._setupStream (client.js:240)
    at new MqttClient (client.js:227)
    at Function.connect (index.js:137)
    at new MqttService (mqtt.service.ts:10)
    at _createClass (core.es5.js:9525)
    at _createProviderInstance$1 (core.es5.js:9500)
    at resolveNgModuleDep (core.es5.js:9485)
    at NgModuleRef_.get (core.es5.js:10577)
    at resolveDep (core.es5.js:11080)
    at createClass (core.es5.js:10933)
    at createDirectiveInstance (core.es5.js:10764)
    at createViewNodes (core.es5.js:12212)
    at createRootView (core.es5.js:12107)
    at callWithDebugContext (core.es5.js:13493)
    at Object.debugCreateRootView [as createRootView] (core.es5.js:12810)
    at ComponentFactory_.create (core.es5.js:9872)
    at ComponentFactoryBoundToModule.create (core.es5.js:3333)
    at ViewContainerRef_.createComponent (core.es5.js:10074)
    at RouterOutlet.activateWith (router.es5.js:5617)
    at RouterOutlet.ngOnInit (router.es5.js:5497)
    at checkAndUpdateDirectiveInline (core.es5.js:10856)
    at checkAndUpdateNodeInline (core.es5.js:12364)
    at checkAndUpdateNode (core.es5.js:12303)
    at debugCheckAndUpdateNode (core.es5.js:13167)
    at debugCheckDirectivesFn (core.es5.js:13108)
    at Object.eval [as updateDirectives] (DashboardComponent.html:49)
    at Object.debugUpdateDirectives [as updateDirectives] (core.es5.js:13093)
    at checkAndUpdateView (core.es5.js:12270)
    at callViewAction (core.es5.js:12638)
    at execComponentViewsAction (core.es5.js:12570)
    at checkAndUpdateView (core.es5.js:12276)
    at callViewAction (core.es5.js:12638)
    at execEmbeddedViewsAction (core.es5.js:12596)
    at checkAndUpdateView (core.es5.js:12271)
    at callViewAction (core.es5.js:12638)
    at execComponentViewsAction (core.es5.js:12570)
    at checkAndUpdateView (core.es5.js:12276)
    at callWithDebugContext (core.es5.js:13493)
    at Object.debugCheckAndUpdateView [as checkAndUpdateView] (core.es5.js:13033)
    at ViewRef_.detectChanges (core.es5.js:10189)
    at core.es5.js:4812
    at Array.forEach (<anonymous>)
    at ApplicationRef_.tick (core.es5.js:4812)
    at core.es5.js:4684
    at ZoneDelegate.invoke (zone.js:391)
    at Object.onInvoke (core.es5.js:3890)
    at ZoneDelegate.invoke (zone.js:390)
    at Zone.run (zone.js:141)
    at NgZone.run (core.es5.js:3821)
    at Object.next (core.es5.js:4684)
    at SafeSubscriber.schedulerFn [as _next] (core.es5.js:3635)
    at SafeSubscriber.__tryOrUnsub (Subscriber.js:238)
    at SafeSubscriber.next (Subscriber.js:185)
    at Subscriber._next (Subscriber.js:125)
    at Subscriber.next (Subscriber.js:89)
    at EventEmitter.Subject.next (Subject.js:55)
    at EventEmitter.emit (core.es5.js:3621)
    at checkStable (core.es5.js:3855)
    at Object.onHasTask (core.es5.js:3903)
    at ZoneDelegate.hasTask (zone.js:444)
    at ZoneDelegate._updateTaskCount (zone.js:464)
    at Zone._updateTaskCount (zone.js:288)
    at Zone.runTask (zone.js:208)
    at drainMicroTaskQueue (zone.js:595)
    at <anonymous>
```

**Comment by FreedCapybara** [comment]
It blows up here inside of `_stream_duplex.js` of `readable-stream`:
```js
function Duplex(options) {
  if (!(this instanceof Duplex))
    return new Duplex(options);

  Readable.call(this, options);
  Writable.call(this, options);

  if (options && options.readable === false)
    this.readable = false;

  if (options && options.writable === false)
    this.writable = false;

  this.allowHalfOpen = true;
  if (options && options.allowHalfOpen === false)
    this.allowHalfOpen = false;

  this.once('end', onend); // <--- this guy
}
```
Thanks for looking into this, by the way!

**Comment by RangerMauve** [comment]
This looks like a bug with readable-stream and your build environment. Could you post your webpack config? Also, are you sure it's actually using the version of readable-stream that you specified? Check your node_modules to see if `mqtt/node_modules` has its own readable-stream.

**Comment by FreedCapybara** [comment]
I know it's more of a readable-stream issue, but it's relevant to this library and I just need some help.

Based on the output of `npm list readable-stream`, I think it's safe to say `mqtt.js` is using 2.3.3:
```
+-- mqtt@2.13.0
| +-- concat-stream@1.6.0
| | `-- readable-stream@2.3.3
| +-- help-me@1.1.0
| | +-- callback-stream@1.1.0
| | | `-- readable-stream@1.1.14            <--- except this one, but help-me probably isn't the most important dependency
| | `-- glob-stream@6.1.0
| |   +-- ordered-read-streams@1.0.1
| |   | `-- readable-stream@2.3.3
| |   `-- readable-stream@2.3.3
| +-- mqtt-packet@5.4.0
| | `-- bl@1.2.1
| |   `-- readable-stream@2.3.3
| +-- readable-stream@2.3.3
| `-- websocket-stream@5.0.1
|   +-- duplexify@3.5.1
|   | `-- readable-stream@2.3.3
|   `-- readable-stream@2.3.3
```
Here's my webpack:
```js
const helpers = require('./helpers');
const webpackMerge = require('webpack-merge'); // used to merge webpack configs
// const webpackMergeDll = webpackMerge.strategy({plugins: 'replace'});
const commonConfig = require('./webpack.common.js'); // the settings that are common to prod and dev
const aws = require('aws-sdk');
const creds = new aws.SharedIniFileCredentials({profile: 'profile'});
aws.config.update({region: 'us-east-1', credentials: creds});
const cloudFormation = new aws.CloudFormation({apiVersion: '2010-05-15'});
const s3 = new aws.S3();

/**
 * Webpack Plugins
 */
const AddAssetHtmlPlugin = require('add-asset-html-webpack-plugin');
const DefinePlugin = require('webpack/lib/DefinePlugin');
const NamedModulesPlugin = require('webpack/lib/NamedModulesPlugin');
const LoaderOptionsPlugin = require('webpack/lib/LoaderOptionsPlugin');
const HotModuleReplacementPlugin = require('webpack/lib/HotModuleReplacementPlugin');

/**
 * Webpack Constants
 */
const ENV_VARS = require('./environments/dev.js');
const ENV = process.env.ENV = process.env.NODE_ENV = 'development';
const HOST = process.env.HOST || 'localhost';
const PORT = process.env.PORT || 3000;
const PUBLIC = process.env.PUBLIC_DEV || HOST + ':' + PORT;
const AOT = process.env.BUILD_AOT || helpers.hasNpmFlag('aot');
const HMR = helpers.hasProcessFlag('hot');
const METADATA = {
  host: HOST,
  port: PORT,
  public: PUBLIC,
  ENV: ENV,
  HMR: HMR,
  AOT: AOT
};

/**
 * Webpack configuration
 *
 * See: http://webpack.github.io/docs/configuration.html#cli
 */
module.exports = function (options) {
  const ENV_VARIABLES = {
    'ENV': JSON.stringify(METADATA.ENV),
    'HMR': METADATA.HMR,
    'process': {
      env: {
        'ENV': JSON.stringify(METADATA.ENV),
        'NODE_ENV': JSON.stringify(METADATA.ENV),
        'HMR': JSON.stringify(METADATA.HMR)
      },
      'version': '"v1231123"'
    }
  };

  var keys = Object.keys(ENV_VARS);
  keys.map(k => {
    ENV_VARIABLES[k] = JSON.stringify(ENV_VARS[k]);
  });
  return cloudFormation.describeStacks({StackName: 'stackname'}).promise()
    .then((data) => {
      const allOutputs = data.Stacks[0].Outputs;
      const apiOutput = allOutputs.filter((output) => { return output.OutputKey === 'ServiceEndpoint'; })[0];
      ENV_VARIABLES.PORTAL_API_ENDPOINT = '\"' + apiOutput.OutputValue + '\"';
      return ENV_VARIABLES
    })
    .then((variables) => {
      variables.randomFolderName = Math.random().toString(36).substring(5);
      return variables;
    })
    .then((variables) => {
    return webpackMerge(commonConfig({env: ENV}), {
        /**
         * Developer tool to enhance debugging
         *
         * See: http://webpack.github.io/docs/configuration.html#devtool
         * See: https://github.com/webpack/docs/wiki/build-performance#sourcemaps
         */
        devtool: 'cheap-module-source-map',

        /**
         * Options affecting the output of the compilation.
         *
         * See: http://webpack.github.io/docs/configuration.html#output
         */
        output: {

          /**
           * The output directory as absolute path (required).
           *
           * See: http://webpack.github.io/docs/configuration.html#output-path
           */
          path: helpers.root('dist' + '\\' + variables.randomFolderName),

          /**
           * Use absolute paths when injecting files
           *
           * See: https://github.com/jantimon/html-webpack-plugin/issues/156
           */
          publicPath: '/',

          /**
           * Specifies the name of each output file on disk.
           * IMPORTANT: You must not specify an absolute path here!
           *
           * See: http://webpack.github.io/docs/configuration.html#output-filename
           */
          filename: '[name].bundle.js',

          /**
           * The filename of the SourceMaps for the JavaScript files.
           * They are inside the output.path directory.
           *
           * See: http://webpack.github.io/docs/configuration.html#output-sourcemapfilename
           */
          sourceMapFilename: '[file].map',

          /** The filename of non-entry chunks as relative path
           * inside the output.path directory.
           *
           * See: http://webpack.github.io/docs/configuration.html#output-chunkfilename
           */
          chunkFilename: '[id].chunk.js',

          library: 'ac_[name]',
          libraryTarget: 'var',
        },

        module: {

          rules: [

            /**
             * Css loader support for *.css files (styles directory only)
             * Loads external css styles into the DOM, supports HMR
             *
             */
            {
              test: /\.css$/,
              use: ['style-loader', 'css-loader'],
              include: [helpers.root('src', 'styles')]
            },

            /**
             * Sass loader support for *.scss files (styles directory only)
             * Loads external sass styles into the DOM, supports HMR
             *
             */
            {
              test: /\.scss$/,
              use: ['style-loader', 'css-loader', 'sass-loader'],
              include: [helpers.root('src', 'styles')]
            },

          ]

        },

        plugins: [

          /**
           * Plugin: DefinePlugin
           * Description: Define free variables.
           * Useful for having development builds with debug logging or adding global constants.
           *
           * Environment helpers
           *
           * See: https://webpack.github.io/docs/list-of-plugins.html#defineplugin
           *
           * NOTE: when adding more properties, make sure you include them in custom-typings.d.ts
           */
          new DefinePlugin(variables),

          /**
           * Plugin LoaderOptionsPlugin (experimental)
           *
           * See: https://gist.github.com/sokra/27b24881210b56bbaff7
           */
          new LoaderOptionsPlugin({
            debug: true,
            options: {

            }
          }),

          new HotModuleReplacementPlugin()
        ],

        /**
         * Webpack Development Server configuration
         * Description: The webpack-dev-server is a little node.js Express server.
         * The server emits information about the compilation state to the client,
         * which reacts to those events.
         *
         * See: https://webpack.github.io/docs/webpack-dev-server.html
         */
        devServer: {
          port: METADATA.port,
          host: METADATA.host,
          hot: METADATA.HMR,
          public: METADATA.public,
          historyApiFallback: true,
          watchOptions: {
            // if you're using Docker you may need this
            // aggregateTimeout: 300,
            // poll: 1000,
            ignored: /node_modules/
          },
        },

        /**
         * Include polyfills or mocks for various node stuff
         * Description: Node configuration
         *
         * See: https://webpack.github.io/docs/configuration.html#node
         */
        node: {
          global: true,
          crypto: 'empty',
          process: true,
          module: false,
          clearImmediate: false,
          setImmediate: false
        }

      });
    })
}
```
Thanks!

**Comment by RangerMauve** [comment]
If you're deploying this to the web, you can't use the `mqtt://` protocol, you need to have your MQTT broker configured to use websockets, and then you can use the `ws://` protocol. Does that seem relevant?

**Comment by FreedCapybara** [comment]
That is relevant, but I get the error no matter what URL I use.

**Comment by RangerMauve** [comment]
@mcollina should be well versed on both. Sorry, but I think that's about all I can do with my current knowledge. :(

Check that your config supports the `browser` field for aliasing files

**Comment by FreedCapybara** [comment]
That's okay, thank you for your help! And for your quick response time - I'm glad I didn't have to wait a day to hear from somebody 👍 

Where do I put the browser field in my webpack? I found an issue where somebody asked for it, and some stuff on [this documentation page](https://webpack.github.io/docs/configuration.html) (specifically resolve.packageAlias), but I'm not really sure what to do with it (I know how to configure loaders and stuff, but otherwise I'm not amazing at Webpack).

**Comment by RangerMauve** [comment]
Well, readable-stream was originally geared for use with browserify, and the introduced a new field in the package.json called `browser` which browserify would use to alias files before they got compiled for the web. 

If the browser field isn't being respected by webpack by default, you should add the [aliases defined in readable-stream](https://github.com/nodejs/readable-stream/blob/master/package.json#L41) manually

**Comment by RangerMauve** [comment]
You should look into [resolve.aliasFields](https://webpack.js.org/configuration/resolve/#resolve-aliasfields) to see if that can be configured to use the `browser` field

**Comment by FreedCapybara** [comment]
Between `module` and `plugins` I added:
```js
resolve: {
  aliasFields: ['browser']
},
```
But I don't think it helped =\

**Comment by RangerMauve** [comment]
That's really strange. I'm out of ideas at this point. Sorry!

**Comment by FreedCapybara** [comment]
Thanks for your help!

Closing this in favor of the readable-stream issue https://github.com/nodejs/readable-stream/issues/315

**Comment by xpluscal** [comment]
I have the exact same issue. @FreedCapybara did you end up solving it?

**Comment by FreedCapybara** [comment]
I think so, but you may have to read this thread https://github.com/nodejs/readable-stream/issues/315

TL;DR I believe it's caused by a few things being undefined in readable-stream.

Unfortunately this was all a long time ago and I don't specifically remember what I was working on or if/how I finally overcame the issue.

Sometimes when I run into dependency issues that the maintainers won't fix, I fork the repo, make the edits, and include the code in the project (in a `lib/` folder or something). Then you may be able to do some magic with package.json to make it use that folder as the source of the package.

Good luck, I hope you find a solution!
